File: oop/03-polymorphism/csharp/TransactionProcessor.cs
/*
 * Demonstrates: Polymorphism - Polymorphic Behaviour
 * Concept: Processor works with any Account type without knowing the specific implementation
 */

namespace FinanceDomain.Polymorphism;

public class TransactionProcessor
{
    public List<string> ProcessBatchWithdrawals(IEnumerable<Account> accounts, decimal amount)
    {
        var results = new List<string>();

        foreach (var account in accounts)
        {
            var result = account.ProcessWithdrawal(amount);
            var status = result.Success ? "SUCCESS" : "FAILED";
            results.Add($"[{account.GetAccountType()}] {account.AccountNumber}: {status} - {result.Message}");
        }

        return results;
    }

    public TransactionResult WithdrawFromAccount(Account account, decimal amount)
    {
        return account.ProcessWithdrawal(amount);
    }
}

File: oop/03-polymorphism/typescript/transactionProcessor.ts
/*
 * Demonstrates: Polymorphism - Polymorphic Behaviour
 * Concept: Processor works with any Account type without knowing the specific implementation
 */

import { Account, TransactionResult } from "./account";

export class TransactionProcessor {
    processBatchWithdrawals(accounts: Account[], amount: number): string[] {
        const results: string[] = [];

        for (const account of accounts) {
            const result = account.processWithdrawal(amount);
            const status = result.success ? "SUCCESS" : "FAILED";
            results.push(`[${account.getAccountType()}] ${account.accountNumber}: ${status} - ${result.message}`);
        }

        return results;
    }

    withdrawFromAccount(account: Account, amount: number): TransactionResult {
        return account.processWithdrawal(amount);
    }
}
